### ISSUE TITLE
DW4SF Not Responding to Commands

### ISSUE THREAD

**Issue by gdavids57** [tabrindle/homebridge-leviton#8]
While my DW15S accessories are responding correctly to commands, my DW4SF is not.  I've tried deleting and re-adding the accessory to no effect.  I can control the DW4SF with both the MyLeviton iPhone app and from the MyLeviton web portal so I know the accessory is operating correctly.  

My configuration is as follows:

OS: Raspbian Buster (10), Kernel version:5.4.83-v7l+
Node.js version: v14.15.4
Npm version: v6.14.10
Homebridge version: v1.1.7
Homebridge-Leviton plugin version: v1.1.0

Log entries
Startup
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] Initializing LevitonDecoraSmart platform...
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] configureAccessory Family Ceiling Fan
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] setupService Family Ceiling Fan
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart]   -Device Model:  DW4SF
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart]   -Setting up device as Fan: Family Ceiling Fan
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] getStatus Family Ceiling Fan
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] configureAccessory Porch Light
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] setupService Porch Light
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart]   -Device Model:  DW15S
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart]   -Setting up device as Switch: Porch Light
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] getStatus Porch Light
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] didFinishLaunching
[2/20/2021, 12:28:32 PM] [LevitonDecoraSmart] initialize

Homekit polling for status
[2/20/2021, 12:20:53 PM] [LevitonDecoraSmart] onGetPower callback undefined
[2/20/2021, 12:20:53 PM] [LevitonDecoraSmart] onGetRotationSpeed callback undefined
[2/20/2021, 12:20:53 PM] [LevitonDecoraSmart] onGetPower callback OFF

attempted to turn on ceiling fan - failed
[2/20/2021, 12:21:18 PM] [LevitonDecoraSmart] onSetPower Family Ceiling Fan true
[2/20/2021, 12:21:18 PM] [LevitonDecoraSmart] onSetPower callback undefined

attempted to turn on porch light - successful
[2/20/2021, 12:25:45 PM] [LevitonDecoraSmart] onSetPower Porch Light true
[2/20/2021, 12:25:46 PM] [LevitonDecoraSmart] onSetPower callback ON

attempted to turn off porch light - successful
[2/20/2021, 12:26:01 PM] [LevitonDecoraSmart] onSetPower Porch Light false
[2/20/2021, 12:26:01 PM] [LevitonDecoraSmart] onSetPower callback OFF

**Comment by tabrindle** [comment]
Yeah I've noticed my devices aren't working properly either, will look into this soon.

**Comment by gdavids57** [comment]
I switched over to HOOBS v3.3.5 and the DW4SF started working.  I don't know enough about the two implementations to even render an opinion but perhaps this fact can help in the troubleshooting.

**Comment by jmacdonagh** [comment]
My DW4SF fan switches are no longer responding via HomeKit. They respond via the My Leviton app.

I see this in the logs when I reboot the HomeBridge server:

> [7/18/2021, 20:09:49] [LevitonDecoraSmart] onGetRotationSpeed Living Room undefined%
> [7/18/2021, 20:09:49] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://git.io/JtMGR for more info.

HomeBridge, plugin, and fan switch firmware are all up-to-date.

**Comment by lucanaut** [comment]
Hello, I just noticed that I am experiencing this issue as well.  Everything was working and then...it wasn't.  When I try to turn on the fan, I get the following too:

[2/4/2022, 9:09:48 AM] [homebridge-leviton] onGetRotationSpeed Ceiling Fan undefined%
[2/4/2022, 9:09:48 AM] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://git.io/JtMGR for more info.
[2/4/2022, 9:09:48 AM] [homebridge-leviton] Error: 
    at RotationSpeed.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2142:105)
    at RotationSpeed.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2030:14)
    at RotationSpeed.Characteristic.updateValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1457:20)
    at /usr/local/lib/node_modules/homebridge-leviton/index.js:168:67
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2/4/2022, 9:09:48 AM] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://git.io/JtMGR for more info.
[2/4/2022, 9:09:48 AM] [homebridge-leviton] Error: 
    at RotationSpeed.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2142:105)
    at RotationSpeed.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2030:14)
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1617:24
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:10:18
    at /usr/local/lib/node_modules/homebridge-leviton/index.js:169:11
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

**Comment by tabrindle** [comment]
I think this was a firmware issue. They've since been updated and I've had no problems. Please open a new issue if you find something else.

**Comment by jmacdonagh** [comment]
Hi @tabrindle, what version of the firmware are you on? All of my DW4SF's are now on `1.5.72; CP 1.19`. I updated the Homebridge Leviton plugin to the latest (`v1.4.0`). I restarted Homebridge and still have the same issues:

```
...
[5/16/2022, 19:18:57] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://homebridge.io/w/JtMGR for more info.
[5/16/2022, 19:18:57] [LevitonDecoraSmart] onGetPower Family Room Fan undefined
[5/16/2022, 19:18:57] [LevitonDecoraSmart] onGetRotationSpeed Family Room Fan undefined%
[5/16/2022, 19:18:57] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://homebridge.io/w/JtMGR for more info.
[5/16/2022, 19:18:57] [homebridge-leviton] This plugin generated a warning from the characteristic 'Rotation Speed': characteristic value expected valid finite number and received "undefined" (undefined). See https://homebridge.io/w/JtMGR for more info.
...
```

**Comment by jmacdonagh** [comment]
If you have these devices and confirm that they're currently working for you, then I can add some `console.log()`s to the plugin source to see what the output of the API requests are. Just wanted to make sure before I start.

**Comment by jmacdonagh** [comment]
Hey @tabrindle, I wanted to give you an update. I fixed this by removing the plugin and re-installing.

I observed that each fan's `onGetPower` was using a different authorization token than what we got back from the login call, and so `onGetPower` was getting a JSON response payload that basically said `{"status" : "401"}` (but of course it returned an HTTP 200 so we never caught it).

I'm not sure how this happened, but I noticed that `onGetPower` is called before the plugin's `initialize` (where we'd log in and get the new authorization token). Is some other authorization token stored away for each accessory?
